src/core/shape_inference/include/reshape_shape_inference.hpp

```diff
@@ -0,0 +1,360 @@
+                    output_shape.push_back(*input_iter);
+                    // Exclude special zero dimension from product calculation
+                } else {
+                    output_shape.push_back(pattern_dim);
+                    product.update_out(pattern_dim);
+                }
+
+                if (input_iter != input_last) {
+                    if (!ignore_pattern_dim) {
+                        product.update_in(*input_iter);
+                    }
+                    ++input_iter;
+                }
+            }
+
+            // update input product by remaining input dimensions.
+            for (; input_iter != input_last; ++input_iter) {
+                product.update_in(*input_iter);
+            }
+        }
+        product.calculate();
+
+        // resolving -1 masked dimension
+        const auto has_minus_one_idx = !dim::is_inf_bound(minus_one_idx);
+        if (has_minus_one_idx) {
+            auto& minus_one_dim = output_shape[minus_one_idx];
+            minus_one_dim = reshape::resolve_minus_one_dim(product);
+
+            if (product.get_static_out() == 0) {
+                NODE_VALIDATION_CHECK(op,
+                                      product.get_static_in() == 0,
+                                      "Cannot infer '-1' dimension with zero-size output dimension unless at least one "
+                                      "input dimension is also zero-size");
+            } else {
+                NODE_VALIDATION_CHECK(op,
+                                      !dim::is_empty(minus_one_dim),
+                                      "Non-'-1' output dimensions do not evenly divide the input dimensions");
+            }
+        }
+
+        if (input_shape.is_static() && output_shape.is_static()) {
+            const auto zero_dims = std::any_of(output_pattern.begin(), output_pattern.end(), cmp::Equal<TDim>(0));
+            const auto backward_compatible_check = (zero_dims && special_zero) || has_minus_one_idx;
+            const auto in_out_elements_equal = (product.get_static_in() == product.get_static_out());
+
+            NODE_SHAPE_INFER_CHECK(op,
+                                   input_shapes,
+                                   backward_compatible_check || in_out_elements_equal,
+                                   "Requested output shape ",
+                                   output_shape,
+                                   " is incompatible with input shape");
+        }
+    } else if (pattern_shape_rank.is_static()) {
+        if (pattern_shape_rank.get_length() == 0) {
+            NODE_SHAPE_INFER_CHECK(op,
+                                   input_shapes,
+                                   input_rank.compatible(0),
+                                   "Input must be scalar as pattern is scalar!");
+        } else {
+            output_shape =
+                PartialShape::dynamic(Rank(pattern_shape[0].get_min_length(), pattern_shape[0].get_max_length()));
```

Comment:
rank is not stored in PartialShape, so you could've coded a bit simpler:
```suggestion
                PartialShape::dynamic(Rank(pattern_shape[0]));
```

Reply:
The `Rank(pattern_shape[0])` cannot be used as there is compile error rank cannot be created from `StaticShape` class.